build/tools/build_parallel.js

```diff
@@ -0,0 +1,203 @@
+
+var path = require('path');
+var fs = require('fs');
+var findRoot = require('find-root');
+var util = require('util');
+var runAll = require('npm-run-all');
+var async = require('async');
+var EventEmitter = require('events');
+
+// See section titled "About MaxListenersExceededWarning" in https://github.com/mysticatea/npm-run-all/blob/HEAD/docs/node-api.md
+EventEmitter.defaultMaxListeners = 30;
+
+var usage = function() {
+  console.log('usage: node ' + __filename + ' task');
+  console.log('where task is one of:');
+  var tasks = require('./projects.json').tasks;
+  Object.keys(tasks).forEach(function(taskName) {
+    var task = tasks[taskName];
+    console.log("  " + taskName + " - " + task.description);
+  });
+;}
+
+var taskName = (process.argv.length > 2 && process.argv[2]) || 'test';
+var task = require('./projects.json').tasks[taskName];
+if (!task) {
+  usage();
+  process.exit(1);
+}
+
+console.log('dirname is ' + __dirname);
+var gitRoot = findRoot(__dirname, function (dir) {
+  return fs.existsSync(path.resolve(dir, '.gitattributes'));
+});
+
+// ingest projects.json
+var readProjectFile = function() {
+  var projectList = require('./projects.json').projects;
+  var projects = {};
+  projectList.forEach(function (project) {
+    project.fullpath = path.join(gitRoot, project.directory);
+    var packageJson = require(path.join(project.fullpath, 'package.json'));
+    project.dependencies = packageJson.dependencies;
+    project.name = packageJson.name;
+    projects[project.name] = project;
+  });
+  return projects;
+};
+var projects = readProjectFile();
+
+// helper to see if we run the current task for a given project`
+var needToRunTask = function(project) {
+  return (!project.hasOwnProperty(taskName) || !!project[taskName]);
+};
+
+// make a temporary package.json to run our scripts
+var tempDir, packageJsonFilename;
+var makePackageJson = function() {
+  tempDir = path.join(__dirname, 'temp');
+  try {
+    fs.statSync(tempDir).isDirectory();
+  } catch (err) {
+    fs.mkdirSync(tempDir);
+  }
+
+  var packageJson = { 'scripts' : {} };
+
+  Object.keys(projects).forEach(function (projectName) {
+    var project = projects[projectName];
+    if (needToRunTask(project)) {
+      var command = task.command;
+      command = command.replace(/{fullpath}/g, project.fullpath);
+      if (task.hasOwnProperty('perDependencyCommand')) {
+        var dependencyCommand = '';
+        Object.keys(project.dependencies).forEach(function(dependencyName) {
+          if (!!projects[dependencyName]) {
+            dependencyCommand += task.perDependencyCommand.replace(/{dependency}/g, dependencyName);
+          }
+        });
+        command = command.replace(/{perDependencyCommand}/g, dependencyCommand);
+      }
+      packageJson.scripts[taskName + '_' + project.name] = command;
+    }
+  });
+  packageJsonFilename = path.join(tempDir, 'package.json');
+  fs.writeFileSync(packageJsonFilename, JSON.stringify(packageJson,null,'  '));
```

Comment:
this whole thing is elegant but extremely hard to understand. I'm afraid maintaining it might become a headache for anyone that is not you.
Please consider at least adding comments, and maybe splitting it in functions that are named after each of the steps you're taking, and also try to have less repetitive names for variables.